python_modules/dagster/dagster/_daemon/monitoring/monitoring_daemon.py

```diff
@@ -102,34 +114,93 @@ def execute_monitoring_iteration(
+        instance.get_run_records(filters=RunsFilter(statuses=IN_PROGRESS_RUN_STATUSES))
+    )
 
-    if not runs:
+    if not run_records:
         return
 
-    logger.info(f"Collected {len(runs)} runs for monitoring")
+    logger.info(f"Collected {len(run_records)} runs for monitoring")
     workspace = workspace_process_context.create_request_context()
-    for run in runs:
+    for run_record in run_records:
         try:
-            logger.info(f"Checking run {run.run_id}")
+            logger.info(f"Checking run {run_record.dagster_run.run_id}")
 
-            if run.status == DagsterRunStatus.STARTING:
-                monitor_starting_run(instance, run, logger)
-            elif run.status == DagsterRunStatus.STARTED:
-                monitor_started_run(instance, workspace, run, logger)
-            elif run.status == DagsterRunStatus.CANCELING:
+            if run_record.dagster_run.status == DagsterRunStatus.STARTING:
+                monitor_starting_run(instance, run_record, logger)
+            elif run_record.dagster_run.status == DagsterRunStatus.STARTED:
+                monitor_started_run(instance, workspace, run_record, logger)
+            elif run_record.dagster_run.status == DagsterRunStatus.CANCELING:
                 # TODO: implement canceling timeouts
                 pass
             else:
-                check.invariant(False, f"Unexpected run status: {run.status}")
+                check.invariant(False, f"Unexpected run status: {run_record.dagster_run.status}")
         except Exception:
             error_info = serializable_error_info_from_exc_info(sys.exc_info())
-            logger.error(f"Hit error while monitoring run {run.run_id}: {str(error_info)}")
+            logger.error(
+                f"Hit error while monitoring run {run_record.dagster_run.run_id}: {str(error_info)}"
+            )
             yield error_info
         else:
             yield
+
+
+def check_run_timeout(
+    instance: DagsterInstance, run_record: RunRecord, logger: logging.Logger
+) -> None:
+    max_time_str = run_record.dagster_run.tags.get(
+        MAX_RUNTIME_SECONDS_TAG,
+    )
+    if not max_time_str:
+        return
+
+    max_time = float(max_time_str)
+
+    if (
+        run_record.start_time is not None
+        and pendulum.now("UTC").timestamp() - run_record.start_time > max_time
+    ):
+        logger.info(
+            f"Run {run_record.dagster_run.run_id} has exceeded maximum runtime of"
+            f" {max_time} seconds: terminating run."
+        )
+
+        instance.report_run_canceling(
+            run_record.dagster_run,
+            message=f"Canceling due to exceeding maximum runtime of {max_time} seconds.",
+        )
+        try:
+            if instance.run_launcher.terminate(run_id=run_record.dagster_run.run_id):
+                instance.report_run_failed(
+                    run_record.dagster_run, f"Exceeded maximum runtime of {max_time} seconds."
+                )
```

Comment:
can you show an event log for a run that goes through this? This is the first time that i'm aware of that we knowingly terminate a run while its in a FAILURE state and i'm curious what the last event that the run worker emits looks like

Reply:
Here's an example of the following job going through the termination flow:
```python
import time

from dagster import job, op, resource


@resource
def some_resource_with_spindown(context):
    try:
        yield "foo"
    finally:
        context.log.info("Spinning down resource")


@op(required_resource_keys={"foo"})
def some_op():
    time.sleep(60)
    return 1



@job(resource_defs={"foo": some_resource_with_spindown})
def the_job():
    some_op()
    some_op()
```
I induced this by enabling run monitoring with a 5 second polling interval, and setting the tag with 5 seconds as the maximum time. The following was the event stream post cancellation:
![Screen Shot 2023-04-05 at 1 50 53 PM](https://user-images.githubusercontent.com/17576880/230208938-409ec5bc-2074-4eb5-a272-ff90fb661b5c.png)
Basically, you end up getting the actual step-related events + spindown stuff after the run is marked as a failure.

Reply:
Ended up changing this to avoid the two `run_failure` events above; now the second event doesn't produce a run failure, but instead an engine event:
![Screen Shot 2023-04-05 at 2 15 08 PM](https://user-images.githubusercontent.com/17576880/230213255-735bce25-4b47-4506-a8db-cf9bd85967cb.png)